### PR TITLE
Bipartite R-mat graph generation.

### PR DESCRIPTION
**Pull request by seunghwak**
Addresses https://github.com/rapidsai/cugraph/issues/2075

This function will generate (source, destination) vertex ID pairs. Source vertex IDs will have values in `[0, 2^src_scale)` and destination vertex IDs will have values in `[0, 2^dst_scale)`.

Additionally,

* `scramble_vertex_ids` function had unused input parameters and it was internally erroneously setting scale. Fixed this bug.

* Rmat_Usecase was ignoring scramble_vertex_ids flag, fixed this bug.

* Added `scramble_vertex_ids` that take a just single vertex list (instead of src, dst pair)

* Update `scramble_vertex_ids` to take input vectors as R-values and return scrambled vectors (instead taking in/out parameters)

### CODE REVIEW

**Comment by seunghwak** [comment]
@BradReesWork Will this API be sufficient to address the issue https://github.com/rapidsai/cugraph/issues/2075?

**Comment by BradReesWork** [comment]
@seunghwak the num_edges  will be the number of edges between the sets?   Also how are you enforcing that edges are only between sets and no within a set?

**Comment by seunghwak** [comment]
> the num_edges will be the number of edges between the sets?
Yes.

> Also how are you enforcing that edges are only between sets and no within a set?

Here, expectation is that source vertices belong to one set and destination vertices belong to the other set.

Say there are two sets U and V.

Say U has 2^u_scale vertices and V has 2^v_scale vertices.

To generate edges from U to V, you call this function with src_scale = u_scale and dst_scale = v_scale.

To generate edges from V to U, you call this function with src_scale = v_scale and dst_scale = u_scale.

And the caller wants vertex IDs for U to be in [0, 2^u_scale) and V to be in [2^u_scale, 2^u_scale + 2^v_scale), they need to add 2^u_scale for either source or vertex IDs after calling this function.

**Review by naimnv — APPROVED** [review]

**Comment by ajschmidt8** [comment]
@seunghwak, just an FYI. adding `[skip-ci]` to the PR title doesn't do anything for GitHub Actions.

please review the page below to figure out how to skip CI for GitHub Actions.

- https://docs.rapids.ai/resources/github-actions/#skipping-ci-for-commits

**Comment by seunghwak** [comment]
> Can the caller have any other options than adding |U| to destination vertexIDs (or adding |V| to source vertexIds) if the output edge-lists are used to create graph using `cugraph::create_graph_from_edgelist`?

What do you mean? To explicitly create a bi-partite graph? We currently do not store bi-partitie graphs in a different format, but in the future and if it turns out to be hugely beneficial, we may consider taking (vertex ID within a set, set number) pairs instead of simple vertex IDs for k-partitie graphs. Not considering this right at this moment.

**Comment by seunghwak** [comment]
@ChuckHastings Can we delete the deprecated R-mat generators (which takes a seed) or do we still need to keep them?

**Comment by ChuckHastings** [comment]
> @ChuckHastings Can we delete the deprecated R-mat generators (which takes a seed) or do we still need to keep them?

I believe we need to keep them until the python code is modified to use the new C API approach.  This work is tracked in #3480

File: cpp/src/generators/generate_rmat_edgelist.cu [hunk 1]
```diff
@@ -212,6 +213,83 @@ generate_rmat_edgelists(raft::handle_t const& handle,
...
+            dst += dst_bit_set ? static_cast<vertex_t>(vertex_t{1} << (dst_scale - (level + 1))) : 0;
+          }
+        }
+        return thrust::make_tuple(src, dst);
+      });
+    num_edges_generated += num_edges_to_generate;
+  }
+
+  return std::make_tuple(std::move(srcs), std::move(dsts));
+}
```
Comment (naimnv) [inline cpp/src/generators/generate_rmat_edgelist.cu:291]: Just wondering, is it the algorithm used here?
https://www.cambridge.org/core/journals/network-science/article/linear-work-generation-of-rmat-graphs/68A0DDA58A7B84E9B3ACA2DBB123A16C

**Review by naimnv — APPROVED** [review]

**Comment by naimnv** [comment]
> What do you mean?

Yeah right, the vertex IDs in U+V has be to be unique.

File: cpp/src/generators/generate_rmat_edgelist.cu (hunk 1 above)
Comment (seunghwak) [reply cpp/src/generators/generate_rmat_edgelist.cu]: Possibly, but this code basically follows the algorithm used in graph 500 reference code.

Comment (seunghwak) [reply cpp/src/generators/generate_rmat_edgelist.cu]: Maybe not... I just skimmed the abstract, and it seems like the paper claims that edge generation time is just function of the number of edges to generate and irrelevant to scale.

Here, it is dependent on scale. Not sure the algorithm in the paper might actually be faster, but here, scale is pretty much limited, and R-mat graph generation is fast enough for our use cases.

**Review by ChuckHastings — APPROVED** [review]

**Review by naimnv — APPROVED** [review]

**Review by BradReesWork — APPROVED** [review]

**Pull request merged**